syfertext/tokenizer.py

```diff
@@ -175,18 +217,411 @@ def __call__(self, text: Union[String, str]):
+        # list of TokenMeta of exceptions after splitting the affixes.
+        substring, pos, affixes, exception_tokens = self._split_affixes(
+            substring=substring, start_pos=pos
+        )
+
+        # Attach all the `TokenMeta` objects formed as result of splitting
+        # affixes in the doc container.
+        doc = self._attach_tokens(
+            doc=doc,
+            substring=substring,
+            start_pos=pos,
+            space_after=space_after,
+            affixes=affixes,
+            exception_tokens=exception_tokens,
+        )
 
         return doc
 
+    def _split_affixes(
+        self, substring: str, start_pos: int
+    ) -> Tuple[str, int, DefaultDict, List[TokenMeta]]:
+        """Process substring for tokenizing prefixes, infixes, suffixes and exceptions.
+
+        Args:
+            substring: The substring to tokenize.
+            start_pos: A pointer to the start position of the substring in the text.
+
+        Returns:    
+            substring: The substring to tokenize.
+            start_pos: A pointer to the start position of the substring in the text.
+            affixes: Dict holding TokenMeta lists of each affix 
+                types as result of splitting affixes
+            exception_tokens: The list of exception tokens TokenMeta objects.
+        """
+        suffixes = []
+        prefixes = []
+        infixes = []
+        exception_tokens = []
+        pos = start_pos
+        end_pos = pos
+
+        next_affix = ["prefix", "suffix"]
+
+        # Dict holding TokenMeta lists of each affix types(prefix, suffix, infix)
+        affixes = defaultdict(list)
+
+        # The first element in the `next_affix` list is 'prefix'.
+        # since we should start by finding prefixes, we fix the
+        # index i = 0.
+        # Start by finding prefixes.
+        i = 0
+
+        # Loop until an empty substring is encountered
+        while self.find_prefix(substring) or self.find_suffix(substring):
+
+            if substring in self.exceptions:
+                break
+
+            # Get affix type (prefix or suffix)
+            affix_type = next_affix[i % 2]
+
+            affix_finder = getattr(self, f"find_{affix_type}")
+
+            if affix_finder(substring):
+                # Get the `TokenMeta` object of the affix along with updated
+                # substring and start pos pointer after removing the affix
+                token_meta, substring, pos = getattr(self, f"_get_{affix_type}_token_meta")(
+                    substring, pos
+                )
+
+                affixes[f"{affix_type}"].append(token_meta)
+
+            # Change the affix type.
+            i += 1
+
+        # Check if substring is in exceptions
+        if substring in self.exceptions:
+            # Get a list of exception  `TokenMeta` objects to be added in the Doc container
+            exception_tokens, substring = self._get_exception_token_metas(substring, pos)
```

Comment:
You no more need this section:

```
# Check if substring is in exceptions
if substring in self.exceptions:
    # Get a list of exception  `TokenMeta` objects to be added in the Doc container
    exception_tokens, substring = self._get_exception_token_metas(substring, pos)
```

Since since you check for exceptions in the loop, right?

Reply:
yeah, i just need to call the _get_exception_token_meta in loop. (previously I just break the loop on that point)